devel/style_guide.rst

```diff
@@ -0,0 +1,144 @@
+
+Commenting
+^^^^^^^^^^
+
+Functions inside of header files should include doxygen-style comments, including documentation for all parameters and return values. Implementation of those methods in `cc` files do not need to include the comment.
+
+Non-obvious algorithms should include comments about what the code is doing to aid in later maintenance. Avoid writing comments for code where it is obvious what that code is doing.
+
+Spaces and control statements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Spaces should exist inside the outer parentheses for all control statements, but not function calls. A space should also follow the keyword. For example:
+
+::
+
+  if ( condition )
+	{
+	}
+	
+  SomeFunction(arg1);
+
+Space after not-operator (`!`)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A space should exist after any not-operator. For example:
+
+::
+
+  if ( ! condition )
+	{
+	}
+
+Pointer and reference arguments/variables
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Pointer and reference characters should modify the type of the argument, not the variable. For example, `int* var` and not `int *var`;
+
+Visibility ordering
+^^^^^^^^^^^^^^^^^^^
+
+Use the ordering `public` -> `protected` -> `private` in class definitions for members. If the class includes `friend` methods, list those at the start of the class.
+
+Language support and preferences
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Zeek build system only supports up to C++11 as supported by gcc 4.8.3 (meaning not the whole C++11 feature set). This is due to compiler support on the target platforms users have reported needing.
+
+Exceptions
+**********
+
+Avoid using exceptions for error handling. The primary reason to avoid them is that it makes error handling more difficult to reason about. Due to the nature of the reference counting in the Zeek code, exceptions will often cause the counting to be invalid unless handled very carefully.
+
+Casting
+*******
+
+Use C++-style casting (`static_cast`, `dynamic_cast`, `reinterpret_cast`, `const_cast`) instead of bare C-style casts.
+
+Strings
+*******
+
+One artifact of the long life of this Zeek code is that a large number of the strings are plain `char*` values. For new code, use std::string instead.
+
+Explicit constructors
+*********************
+
+Single-argument constructors should be marked `explicit` to aid in type-checking.
+
+Global namespace
+****************
+
+Another artifact of the old Zeek code is that a large amount of variables, functions, and constants are defined in the global namespace and then `extern`'d when needed in other places. Avoid adding any more to the global namespace when possible. Prefer using constructs like the Singleton pattern or static class members instead.
+
+Function parameter passing
+**************************
+
+Follow the typical C++ best practices for parameter passing. Never pass objects by value, instead passing them by reference or const reference depending on if they're going to be modified.
```

Comment:
"Never" feels a bit much.  Maybe more like "If your function can make use of move semantics, then fine to use a by-value parameter where the caller may decide to use an `std::move()` argument or not.  In other cases, use a (const-)reference".

For passing non-const-reference versus pointer in cases where you want to mutate the argument, I sometimes prefer to use a pointer because call-sites can then be used to help distinguish what may change.  e.g. in `foo(&bar)` versus `foo(bar)`, I'm more likely to suspect the former to modify `bar`.  Don't care much about formalizing that as standard convention, just maybe like the idea of not being too strict here and leaving some up to judgement of the programmer.

Reply:
The pass-by-pointer-to-modify style is something I like, too. Or put differently: I don't like passing by non-const reference, because it's not clear what'll happen to the instance.

Reply:
How about this:

> Follow the typical C++ best practices for parameter passing. Avoid passing large objects by value, except in cases where the function can use move semantics and the caller can use `std::move`. For objects that will not be modified by the function, pass by const-reference. For objects that may be modified by the function, prefer making the argument a pointer.